pd-client/client.go

```diff
@@ -43,77 +50,271 @@ type Client interface {
+
+	c.wg.Add(2)
+	go c.tsLoop()
+	go c.leaderLoop()
+
+	// TODO: Update addrs from server continuously by using GetMember.
+
+	return c, nil
+}
+
+func (c *client) initClusterID() error {
+	for i := 0; i < maxInitClusterRetries; i++ {
+		for _, u := range c.urls {
+			client, err := apiutil.NewClient(u, pdTimeout)
+			if err != nil {
+				log.Errorf("[pd] failed to get cluster id: %v", err)
+				continue
+			}
+			clusterID, err := client.GetClusterID()
+			if err != nil {
+				log.Errorf("[pd] failed to get cluster id: %v", err)
+				continue
+			}
+			c.clusterID = clusterID
+			return nil
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	return errors.New("failed to get cluster id")
+}
+
+func (c *client) updateLeader() error {
+	for _, u := range c.urls {
+		client, err := apiutil.NewClient(u, pdTimeout)
+		if err != nil {
+			continue
+		}
+		leader, err := client.GetLeader()
+		if err != nil {
+			continue
+		}
+		c.mu.RLock()
+		changed := c.leader != leader.GetAddr()
+		c.mu.RUnlock()
+		if changed {
+			if err = c.switchLeader(leader.GetAddr()); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	}
+	return errors.Errorf("failed to get leader from %v", c.urls)
+}
+
+func (c *client) switchLeader(addr string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.clients[addr]; !ok {
+		cc, err := grpc.Dial(addr, grpc.WithDialer(func(addr string, d time.Duration) (net.Conn, error) {
+			u, err := url.Parse(addr)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			return net.Dial(u.Scheme, u.Host)
+		}), grpc.WithInsecure())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.clients[addr] = pdpb2.NewPDClient(cc)
+	}
+	c.leader = addr
+	return nil
+}
+
+func (c *client) leaderLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case <-c.checkLeaderCh:
+		case <-time.After(time.Minute):
+		case <-c.quit:
+			return
+		}
+
+		if err := c.updateLeader(); err != nil {
+			log.Errorf("[pd] failed updateLeader: %v", err)
+		}
+	}
+}
+
+func (c *client) tsLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case firstReq := <-c.tsoRequests:
+			start := time.Now()
+			ctx, cancel := context.WithTimeout(context.Background(), pdTimeout)
+			defer cancel()
+
+			pendingCount := len(c.tsoRequests)
+			resp, err := c.leaderClient().Tso(ctx, &pdpb2.TsoRequest{
+				Header: &pdpb2.RequestHeader{ClusterId: c.clusterID},
+				Count:  uint32(pendingCount + 1),
+			})
+			requestDuration.WithLabelValues("tso").Observe(time.Since(start).Seconds())
+			if err == nil && resp.GetCount() != uint32(pendingCount+1) {
+				err = errors.New("[pd] tso length in rpc response is incorrect")
+			}
+			if err != nil {
+				firstReq.done <- errors.Trace(err)
+				for i := 0; i < pendingCount; i++ {
+					req := <-c.tsoRequests
+					req.done <- errors.Trace(err)
+				}
+			}
+			tsHigh := resp.GetTimestamp()
+			firstReq.physical = tsHigh.GetPhysical()
+			firstReq.logical = tsHigh.GetLogical() - int64(pendingCount)
```

Comment:
you can calculate the first logical, and then use this later. no need to do `tsHigh.GetLogical() - int64(resp.GetCount())` in every loop.